File: src/examples/javascript/merge-sort.js
/**
 * Merge Sort - JavaScript
 *
 * Divide-and-conquer sorting algorithm.
 * Recursively splits the array in half, sorts each half, then merges.
 *
 * Time Complexity: O(n log n)
 * Space Complexity: O(n)
 */
export default {
  id: 'merge-sort-js',
  language: 'javascript',
  title: 'Merge Sort',
  description: 'Divide-and-conquer sorting with recursive merging',
  difficulty: 'intermediate',
  category: 'sorting',
  tags: ['algorithm', 'sorting', 'recursion', 'divide-and-conquer'],
  code: `function mergeSort(arr) {
  if (arr.length <= 1) return arr;

  const mid = Math.floor(arr.length / 2);
  const left = mergeSort(arr.slice(0, mid));
  const right = mergeSort(arr.slice(mid));

  return merge(left, right);
}

function merge(left, right) {
  const result = [];
  let i = 0, j = 0;

  while (i < left.length && j < right.length) {
    if (left[i] <= right[j]) {
      result.push(left[i]);
      i++;
    } else {
      result.push(right[j]);
      j++;
    }
  }

  return result.concat(left.slice(i)).concat(right.slice(j));
}

const nums = [38, 27, 43, 3, 9, 82, 10];
console.log("Original:", nums);
const sorted = mergeSort(nums);
console.log("Sorted:", sorted);`,
};

File: src/examples/python/merge-sort.py
"""
Merge Sort - Python

Efficient divide-and-conquer sorting algorithm.
Divides array into halves, sorts them, and merges.

Time Complexity: O(n log n)
Space Complexity: O(n)
"""

def merge_sort(arr):
    """
    Sort array using merge sort algorithm.
    
    Args:
        arr: List of comparable elements
        
    Returns:
        Sorted list
    """
    if len(arr) <= 1:
        return arr
    
    # Divide
    mid = len(arr) // 2
    left = merge_sort(arr[:mid])
    right = merge_sort(arr[mid:])
    
    # Conquer (merge)
    return merge(left, right)

def merge(left, right):
    """Merge two sorted arrays."""
    result = []
    i = j = 0
    
    while i < len(left) and j < len(right):
        if left[i] <= right[j]:
            result.append(left[i])
            i += 1
        else:
            result.append(right[j])
            j += 1
    
    result.extend(left[i:])
    result.extend(right[j:])
    return result

# Test the function
numbers = [64, 34, 25, 12, 22, 11, 90]
print(f"Original: {numbers}")
sorted_nums = merge_sort(numbers)
print(f"Sorted: {sorted_nums}")
